File: src/chatbot_pipeline/documents.py
from __future__ import annotations

import hashlib
import shutil
from pathlib import Path

from PIL import Image, ImageSequence

PDF_EXTENSIONS = {".pdf"}
IMAGE_EXTENSIONS = {".png", ".jpg", ".jpeg", ".webp", ".bmp", ".tif", ".tiff"}
SUPPORTED_EXTENSIONS = PDF_EXTENSIONS | IMAGE_EXTENSIONS


class UnsupportedDocumentError(ValueError):
    pass


def assert_supported(path: Path) -> None:
    if path.suffix.lower() not in SUPPORTED_EXTENSIONS:
        supported = ", ".join(sorted(SUPPORTED_EXTENSIONS))
        raise UnsupportedDocumentError(f"Unsupported input type '{path.suffix}'. Use one of: {supported}")


def stable_file_name(path: Path) -> str:
    digest = hashlib.sha256(path.read_bytes()).hexdigest()[:12]
    safe_stem = "".join(c if c.isalnum() or c in {"-", "_"} else "_" for c in path.stem)
    return f"{safe_stem}-{digest}{path.suffix.lower()}"


def copy_to_uploads(path: Path, upload_dir: Path) -> Path:
    path = path.resolve()
    if not path.exists():
        raise FileNotFoundError(path)
    assert_supported(path)
    target = upload_dir / stable_file_name(path)
    if not target.exists():
        shutil.copy2(path, target)
    return target


def prepare_for_pageindex(path: Path, converted_dir: Path) -> Path:
    """Return a PDF path ready for PageIndex.

    PageIndex's public document-processing SDK currently accepts PDFs. Image inputs,
    including graph screenshots, are converted into one-page or multi-page PDFs first.
    """
    path = path.resolve()
    assert_supported(path)
    if path.suffix.lower() in PDF_EXTENSIONS:
        return path
    return image_to_pdf(path, converted_dir)


def image_to_pdf(path: Path, converted_dir: Path) -> Path:
    converted_dir.mkdir(parents=True, exist_ok=True)
    pdf_path = converted_dir / f"{path.stem}-{hashlib.sha256(path.read_bytes()).hexdigest()[:12]}.pdf"
    if pdf_path.exists():
        return pdf_path

    with Image.open(path) as image:
        frames = [_to_rgb(frame.copy()) for frame in ImageSequence.Iterator(image)]
        if not frames:
            raise UnsupportedDocumentError(f"No readable image frames found in {path}")
        first, *rest = frames
        first.save(pdf_path, "PDF", resolution=150.0, save_all=True, append_images=rest)
    return pdf_path


def _to_rgb(image: Image.Image) -> Image.Image:
    if image.mode == "RGB":
        return image
    if image.mode in {"RGBA", "LA"} or "transparency" in image.info:
        background = Image.new("RGB", image.size, "white")
        alpha = image.convert("RGBA").split()[-1]
        background.paste(image.convert("RGB"), mask=alpha)
        return background
    return image.convert("RGB")

File: src/documents.ts
import { copyFile, mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import { createHash } from "node:crypto";

import { PDFDocument } from "pdf-lib";
import sharp from "sharp";

export const PDF_EXTENSIONS = new Set([".pdf"]);
export const IMAGE_EXTENSIONS = new Set([".png", ".jpg", ".jpeg", ".webp", ".bmp", ".tif", ".tiff"]);
export const SUPPORTED_EXTENSIONS = new Set([...PDF_EXTENSIONS, ...IMAGE_EXTENSIONS]);

export class UnsupportedDocumentError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UnsupportedDocumentError";
  }
}

export function assertSupported(filePath: string): void {
  const extension = path.extname(filePath).toLowerCase();
  if (!SUPPORTED_EXTENSIONS.has(extension)) {
    throw new UnsupportedDocumentError(
      `Unsupported input type '${extension}'. Use one of: ${Array.from(SUPPORTED_EXTENSIONS).sort().join(", ")}`
    );
  }
}

export async function stableFileName(filePath: string, originalName?: string): Promise<string> {
  const sourceName = originalName ?? path.basename(filePath);
  const parsed = path.parse(sourceName);
  const file = await readFile(filePath);
  const digest = createHash("sha256").update(file).digest("hex").slice(0, 12);
  const safeStem = parsed.name.replace(/[^a-z0-9_-]/gi, "_");
  return `${safeStem}-${digest}${parsed.ext.toLowerCase()}`;
}

export async function copyToUploads(
  filePath: string,
  uploadDir: string,
  originalName?: string
): Promise<string> {
  assertSupported(originalName ?? filePath);
  await mkdir(uploadDir, { recursive: true });
  const target = path.join(uploadDir, await stableFileName(filePath, originalName));
  if (path.resolve(filePath) !== path.resolve(target)) {
    await copyFile(filePath, target);
  }
  return target;
}

export async function prepareForPageIndex(filePath: string, convertedDir: string): Promise<string> {
  assertSupported(filePath);
  if (PDF_EXTENSIONS.has(path.extname(filePath).toLowerCase())) {
    return filePath;
  }
  return imageToPdf(filePath, convertedDir);
}

export async function imageToPdf(filePath: string, convertedDir: string): Promise<string> {
  await mkdir(convertedDir, { recursive: true });
  const file = await readFile(filePath);
  const digest = createHash("sha256").update(file).digest("hex").slice(0, 12);
  const pdfPath = path.join(convertedDir, `${path.parse(filePath).name}-${digest}.pdf`);

  const image = sharp(file, { animated: false }).rotate().flatten({ background: "#ffffff" });
  const metadata = await image.metadata();
  const png = await image.png().toBuffer();
  const width = metadata.width ?? 1;
  const height = metadata.height ?? 1;

  const pdf = await PDFDocument.create();
  const page = pdf.addPage([width, height]);
  const embeddedImage = await pdf.embedPng(png);
  page.drawImage(embeddedImage, { x: 0, y: 0, width, height });
  const pdfBytes = await pdf.save();
  await writeFile(pdfPath, pdfBytes);
  return pdfPath;
}
